### ISSUE TITLE
Cloud Based API Instance for Developers

### ISSUE THREAD

**Issue by palisadoes** [PalisadoesFoundation/talawa-api#1428]
We need to have a cloud based API instance for developers to test against. It must:

1. Be integrated into our CI/CD pipeline so that the app is rebuilt with every newly merged PR.
1. Initialize the database to a known working state with appropriate services will be restarted when this happens: (there are existing API calls for this):
    1. with each CI/CD update
    1. every 24 hours
1. Use a talawa.io sub-domain as the endpoint
1. The API instance will have a rate limiting mechanism to reduce the risk of abuse 
1. Preferably use a free service
1. Use a palisadoes role based account for management
1. The demo cloud API must be easily switched between branches as the code source. We will be migrating to master as the default branch soon.

Please coordinate these efforts with @noman2002 and @kb-0311. Ask to be assigned this task by them.

**Comment by kirtanchandak** [comment]
@noman2002 @kb-0311 I can give this issue a try

**Comment by kirtanchandak** [comment]
there is file named ```push-documentation.yml``` in ```github/workflows``` folder. Do we need to make changes here or create a new file?

**Comment by palisadoes** [comment]
1. We used to have this working with Heroku using this file. https://github.com/PalisadoesFoundation/talawa-api/blob/develop/.github/workflows/ci.yml.archive
2. There was some setup that was required with the Heroku cloud to make it work correctly whenever we did a merge. We decided to stop using Heroku due to the cost of having the instance running. We would like to use a fremium service.

**Comment by noman2002** [comment]
@kirtanchandak Please proceed with this. You can use that file, you just need to do some minor changes. But I would suggest please check other solutions like render.com instead of Heroku.

**Comment by vasujain275** [comment]
Hi @noman2002 @kirtanchandak, I have been working on this same issue containerization talawa-api using docker here - https://github.com/PalisadoesFoundation/talawa-api/pull/1418 that was merged yesterday and was working on developing a ci/cd pipeline for this exact use case, @kirtanchandak can I work on this issue if you haven't made any significant progress yet?

**Comment by palisadoes** [comment]
One issue per person

**Comment by vasujain275** [comment]
> One issue per person

Okay

**Comment by palisadoes** [comment]
@kirtanchandak Are you still working on this?

**Comment by vasujain275** [comment]
@palisadoes I can work on this if @kirtanchandak is not working on this

**Comment by palisadoes** [comment]
Assigning to@vasujain275. Please work with @noman2002 and @kb-0311 for comments, advice and reviewing the PR

**Comment by vasujain275** [comment]
Hi @noman2002 and @kb-0311, I did some research and found that the MongoDB Atlas free tier is good enough for our use case. I looked into redis free tier at render.com, it has 25 MB of memory with a 50 connection limit and no persistence storage. We can also host our API on render.com but it will be shut down after an inactivity of 5 minutes and will be restarted when called again. Will that render.com's inactivity delay or redis free tier will be a problem?

**Comment by kb-0311** [comment]
@vasujain275 
1. Instead of render are there other free tier cloud instances that we can use to deploy our application?
2. Well in production the Redis server was to be used as an in-memory cache, not a global cache so we also need to have the option to access the file system/ services of the server where the API is deployed. 
3. Render and stuff is fine but if possible can you also explore free credit use of GCP and/or Azure? Do let me know if those are feasible.

**Comment by vasujain275** [comment]
@kb-0311 We can use the AWS free tier, which will have a  t2.micro ec2 instance free for 12 months per account ( we will need a credit card to avail this 12-month free tier ). We can use the docker file and docker-compose file I created in a recent issue, we can set a ci/cd pipeline to push our image to the docker hub on each push and then pull to e2c using GitHub actions. We can then run docker-compose on e2c also using GitHub ci/cd that can run containers of both redis and MongoDB, although I think it will be best to run MongoDB on Atlas as it has a good free tier and will reduce the load on our t2.micro ec2 instance if we run it on atlas. We can also use azure/gcp, they also have this exact free tier of 12 months that requires a credit card on signup.

**Comment by palisadoes** [comment]
@vasujain275 Ideally we also want:

1. A single account to manage all the required resources. 
1. The ability to have trusted contributors update the cloud service parameters to make sure the CI/CD functions
1. Ensure that only one person has the right to purchase new services or upgrade any existing ones. This will help reduce long term costs and potential abuse.

Other:

1. @kb-0311 @noman2002 Do you have anything else to add?
1. Is the most recent comment from @vasujain275 in keeping with the desired outcome?

**Comment by vasujain275** [comment]
@palisadoes 

1. Only a single aws account is required to manage this cloud instance.
2. We only need to add the cloud service parameters as github secrets of this talawa-api repo. Then the secrets can be safely and easily used in our github actions using environment variables. 
3. Addition of github secrets can be done by any trusted contributors.
4. Only the person who will sign up for aws instance will have the right to purchase new services or upgrade any existing ones. 
5. No one will need to directly ssh into the instance to make changes, as they will be done via ci/cd pipeline.
6. As for MongoDB Atlas, we will only need to add the MongoDB Atlas Cluster url as github secret of this repo.

**Q.** How can github actions do all the changes in our aws instance?
**Ans.** **Docker Containers**, we will never install the talwa api directly to our aws instance instead we will use docker container of our api that can be rebuilt and restarted every time we push changes to our repo.

### My proposed solution - 
1. Firstly, I will create all github actions files and docker compose files required to completely set up talawa-api instance on aws via ci/cd.
2. I will then test all the github actions and compose files on my personal aws ec2 instance to make sure all the things are working as intended. This way I will not be required to have direct access to talawa-api actual production ec2 instance.
3. When all the github actions are completed and tested by me then one of the trusted contributors can signup for aws instance and MongoDB Atlas. Then he will set up the required env variables as github secrets in this repo.

@noman2002 @kb-0311 , appreciate your feedback on possible improvements in this approach

**Comment by palisadoes** [comment]
1. Based on your answer we will need to manage two cloud accounts. Can this be done using only one account?
2. We need to make sure that more than one person can access the AWS dashboard for purposes of troubleshooting. Can this be done without them having access to the credit card information to add new services?
3. If we use a MongoDB Atlas account, how can we guarantee the same AWS access requirement above?

**Comment by kb-0311** [comment]
@vasujain275 
1. Yes your solution and flow make sense to me.
2. At this stage it would be easier to showcase the CI/CD pipeline using your forked talawa-api instance as the base for now and your free tier AWS ec2 instance as the deployment server.
3. Create a new branch and configure the docker-compose file as necessary, since you are the owner of the forked repo you can also add the GitHub secrets for Altas and aws.
4. Then make demo PR merges to that branch and demonstrate how to CI/CD pipeline will work using a video.
5. This will help us understand potential problems that could occur at a given point more precisely.

@palisadoes 
1. Can you elaborate on the two accounts? Are you referring to one account for aws and another for MongoDB Atlas or something else?
2. In aws and MongoDB I think there are certain ABAC and RBAC features for this purpose. Maybe we can take a look at that? @vasujain275

**Comment by vasujain275** [comment]
@palisadoes  Okay I will keep that in mind

**Comment by vasujain275** [comment]
@palisadoes 

Sorry for the delay, I have programmed deploy.yml locally but having some issues with my credit card on aws so not able to test it. I will positively resolve it within a couple of days.

**Comment by palisadoes** [comment]
@vasujain275 If AWS is a difficulty, what would be the minimal Ubuntu VPS required with GoDaddy to get a test instance running? We could provide access via a public/private SSH key for security. 

We really need to get this going.

**Comment by vasujain275** [comment]
@palisadoes
I got the AWS ec2 instance working, I only had debit cards and it was not accepting them. I found a credit card and it worked, I have been testing the code for the past couple of hours. Creating a PR in couple of minutes

**Comment by palisadoes** [comment]
Please answer my question on the GoDaddy VPS.

With that solution we would have a fixed cost and not have to worry about unexpected cloud charges.

Let me know

**Comment by vasujain275** [comment]
@palisadoes 

1. Yeah, we can do a basic GoDaddy VPS and that works fine because we just need an Ubuntu Server. 
2. As for the aws free tier machine, it is really slow and can cause issues in the long run. Build times are around 15min on that.
3. GoDaddy VPS would be a much better option than aws free tier as we don't have to worry about unexpected web charges as you mentioned and will have better performance.
4. Regarding MongoDB hosted on different cloud, I agree with you that having one cloud is better option and from my testing it will not affect our performance. We should go with one cloud instance only.
5. I am improving the docker compose file with more layer caching so that we can reduce build time and overall make the process more efficient.

**Comment by palisadoes** [comment]
If we had a VPS we could host mongo on the local server for simplicity.

What would your minimum be?

- https://www.godaddy.com/hosting/vps-hosting

Remember, we will be resetting the DB every 24 hours. So we don't need a lot of disk.

**Comment by vasujain275** [comment]
@palisadoes
I believe that hosting MongoDB, Redis, and Talawa API on the same VPS is feasible. In this scenario, I think the basic plan (1 vCPU / 2GB RAM / 40GB Storage) would be sufficient.

**Comment by palisadoes** [comment]
@vasujain275

1. Please DM me your ssh public key in slack so I can add you to the server.
2. What directory structure would you propose to get this set up?
3. Under what user would this all run under? I'd prefer for everything to run as a dedicated unprivileged account.

**Comment by vasujain275** [comment]
@palisadoes 

1. Dmed you my ssh public key on slack.
2. We will need only one directory that is talawa-api/ repo dir that will pull new changes on new push. Other data like MongoDB and redis persistent storage will be stored in docker volumes that are created by docker-compose.
3. For routine operations, let's use a dedicated unprivileged account. But for tasks like installing dependencies and GitHub Actions also need sudo. I propose a user with sudo privileges for these tasks.

**Comment by palisadoes** [comment]
1. Check slack for details.
2. Please continue the discussion in the PR
    - https://github.com/PalisadoesFoundation/talawa-api/pull/1491

**Comment by vasujain275** [comment]
@kb-0311 @palisadoes 
So moving forward we need - 

1. All scripts converted to python instead of bash.
2. A way to populate .env file variables from github repo secrets on every merge during deploy.yml workflow.

**Comment by palisadoes** [comment]
Yes

**Comment by palisadoes** [comment]
@vasujain275 @kb-0311 

The workflow is failing. 

- https://github.com/PalisadoesFoundation/talawa-api/actions/runs/7519534011/job/20468173214
- https://github.com/PalisadoesFoundation/talawa-api/actions/runs/7519469484/job/20468033090

This needs to be fixed. It states there is no space on the system, this could also be an authorization issue as I have seen erroneous space related errors related to access.

**Comment by vasujain275** [comment]
@palisadoes 

The cloud instance of talawa-api was out of storage due to old containers and images created by workflow: 
![Screenshot_14-Jan_22-51-38_5075](https://github.com/PalisadoesFoundation/talawa-api/assets/69643310/1622831b-c956-41fa-8b40-c41db76d3828)

Added the step to remove all the unwanted images and containers after the talawa-api containers are up and running in deploy.yaml workflow in this pr - https://github.com/PalisadoesFoundation/talawa-api/pull/1687

Already fixed the deploy.sh file on the GoDaay VPS and tested the workflow too.

**Comment by vasujain275** [comment]
> Do not put the checks in the YAML. This may be another PR

@palisadoes 
Should I only add python scripts to pr for now and leave deploy.yaml workflow unchanged? Please clarify.

**Comment by palisadoes** [comment]
Focus on the scripts first.

**Comment by palisadoes** [comment]
We will need to move the functionality of `deploy.yaml` over to `push.yaml` where all our other push logic is located.

**Comment by palisadoes** [comment]
@vasujain275 

The workflow is failing again:
- https://github.com/PalisadoesFoundation/talawa-api/actions/runs/7543872335/job/20535868792

**Comment by vasujain275** [comment]
@palisadoes 
It failed because I was testing python scripts on the vps, otherwise it is working fine

**Comment by palisadoes** [comment]
1. Thanks @vasujain275 - One more step to go.
    1. https://github.com/PalisadoesFoundation/talawa-api/pull/1704
2. Now we need to move the `deploy.yml` functionality over to `push.yml`

**Comment by vasujain275** [comment]
@palisadoes

1. I am working on moving the functionality from `deploy.yml` to `push.yml`.
2. Additionally, don't we need a way to populate the .env file on the VPS with Talawa API GitHub repository secrets for better security?"

**Comment by palisadoes** [comment]
Yes, that can be done with a separate python deployment script just for creating the .env file. It can
1. Create its own base .env file
1. Use `argparse` for each required variable retrieved from the appropriate GitHub secrets

**Comment by vasujain275** [comment]
@palisadoes The deploy action to Cloud API will be down for a few hours as I am migrating it to a new process and python scripts. I will raise a PR once I am done testing the new scripts thoroughly.

**Comment by Manik2708** [comment]
> @palisadoes The deploy action to Cloud API will be down for a few hours as I am migrating it to a new process and python scripts. I will raise a PR once I am done testing the new scripts thoroughly.

Hi, actually I saw that there are no tests for Redis Caching Services. I have an idea to do this, @palisadoes told me to discuss it with you:
1) We would not do unit tests as mocking can bypass the tests easily.
2) Rather we can use integration tests in which we can directly contact with a testing client which would be running on a different port.
3) Dependency injection can be done with `tsyringe`
Your insights would be helpful!

**Comment by vasujain275** [comment]
@Manik2708, great idea, but we need a better rate limiter for our API right now. I have thought about implementing it with Redis. Currently, I am working on two different issues. Maybe you can first implement the rate limiting using Redis. After that, we can test Redis caching, as the rate limiter is of higher priority and will be beneficial to the project a lot.

**Comment by Manik2708** [comment]
> @Manik2708, great idea, but we need a better rate limiter for our API right now. I have thought about implementing it with Redis. Currently, I am working on two different issues. Maybe you can first implement the rate limiting using Redis. After that, we can test Redis caching, as the rate limiter is of higher priority and will be beneficial to the project a lot.

Is the issue opened for this? Like can I get more context about rate limiting?

**Comment by Manik2708** [comment]
> > @Manik2708, great idea, but we need a better rate limiter for our API right now. I have thought about implementing it with Redis. Currently, I am working on two different issues. Maybe you can first implement the rate limiting using Redis. After that, we can test Redis caching, as the rate limiter is of higher priority and will be beneficial to the project a lot.
> 
> Is the issue opened for this? Like can I get more context about rate limiting?

Also, I saw that Dependency Injection through `tsyringe` doesn't provide control over file execution. For better and strong architecture we have to use `Nest.js` I also raised the issue regarding this in their library: https://github.com/microsoft/tsyringe/issues/235, So there is still a problem of Dependency Injection!

**Comment by vasujain275** [comment]
> > @Manik2708, great idea, but we need a better rate limiter for our API right now. I have thought about implementing it with Redis. Currently, I am working on two different issues. Maybe you can first implement the rate limiting using Redis. After that, we can test Redis caching, as the rate limiter is of higher priority and will be beneficial to the project a lot.
> 
> Is the issue opened for this? Like can I get more context about rate limiting?


1. You need to create a new issue. I will be happy to assist you in creating one if you wish.
2. Currently, we use `express-rate-limiter` to rate limit our API, but it is not an ideal solution for the rate-limiting problem, especially considering that we are soon having a cloud instance for our API.
3. A better approach would be to implement rate limiting by storing IPs as key-value pairs in Redis and limiting access based on that method.
4. Here is the resource that you can refer to:  https://youtu.be/_DqfiW08HkA?si=mZ6wblV3SNPiN0Ru
5. After creating the issue, please comment "Can I work on this issue?" so that maintainers can assign it.

Feel free to mention me in the comments on that issue if you need my help.

**Comment by Manik2708** [comment]
> > > @Manik2708, great idea, but we need a better rate limiter for our API right now. I have thought about implementing it with Redis. Currently, I am working on two different issues. Maybe you can first implement the rate limiting using Redis. After that, we can test Redis caching, as the rate limiter is of higher priority and will be beneficial to the project a lot.
> > 
> > 
> > Is the issue opened for this? Like can I get more context about rate limiting?
> 
> 1. You need to create a new issue. I will be happy to assist you in creating one if you wish.
> 2. Currently, we use `express-rate-limiter` to rate limit our API, but it is not an ideal solution for the rate-limiting problem, especially considering that we are soon having a cloud instance for our API.
> 3. A better approach would be to implement rate limiting by storing IPs as key-value pairs in Redis and limiting access based on that method.
> 4. Here is the resource that you can refer to:  https://youtu.be/_DqfiW08HkA?si=mZ6wblV3SNPiN0Ru
> 5. After creating the issue, please comment "Can I work on this issue?" so that maintainers can assign it.
> 
> Feel free to mention me in the comments on that issue if you need my help.

Thanks for this, surely going to open the issue!

**Comment by vasujain275** [comment]
@Manik2708 So I think we should first let `tsyringe` fix that issue and then proceed with this here

**Comment by palisadoes** [comment]
It's failing. Please resubmit
- https://github.com/PalisadoesFoundation/talawa-api/actions/runs/7769544772

**Comment by vasujain275** [comment]
@palisadoes 
1. It is failing because two commits were added in a very short time gap. 
2. That failed the workflow for latest commit as the previous commit workflow was still running.
3. The api instance is working here correctly - https://api-demo.talawa.io/ 
4. The failed workflow will not affect the deployment of the previous commit to our cloud instance.
5. This edge case only happens when commits are pushed at a small time gap.

**Comment by palisadoes** [comment]
@vasujain275 @Manik2708 

1. Can we let people know this exists and close?
2. Or should we wait for this to be completed?
    1. https://github.com/PalisadoesFoundation/talawa-api/issues/1782

**Comment by vasujain275** [comment]
@palisadoes 

I also think we should wait for the rate limiter issue to complete then we can find an approach to test redis caching.

**Comment by vasujain275** [comment]
@palisadoes 
I have been analyzing the Deploy Workflow for the past couple of days and here are the issues I found - 
1. The workflow will sort of break if there are 2 commits in a very short interval of time. However, it does not affect the functioning of the cloud instance, i.e., no downtime because of this issue.
2. It takes 10-20 min on every new commit to create new docker containers and delete the previous ones. This is **Downtime** for API Instance.

Possible Solution for 2nd Issue - 
1. There will be two workflows - 
    - Firstly, a workflow will create Docker Image of latest state of Talawa API and push it to Docker Hub.
    - Second Workflow will shh into the VPS and simply pull the Docker Image of Talawa API along with other required images like Redis and mongo.
2. This will significantly decrease the time to create new containers, as we only need to pull the prebuilt docker images from the docker hub.


Possible Solution for 1st Issue - 
1. As mentioned in previous solution, if push Docker Image of Talawa API to Docker Hub on every commit. We can only deploy the latest Docker Image every 1hr or so. 
2. This way, for example, If there are 5 commits in that 1hr, only the latest commit's image will be Deployed to the Cloud Instance.

**Comment by kb-0311** [comment]
@vasujain275 Why does the workflow break when the two commits are made in a very short interval of time? Is it because of the remote execution of shell commands in short intervals that is causing the issue or something else entirely?

Agreed on creating two workflows to deploy and pull from the docker hub.

I am still not convinced about the solution to submit only the latest commit at regular time intervals. If there are urgent bug fixes required in production then the amount of time for the bug fix PRs to reflect in prod env would be considerable. 

I think I want to understand the root of the problem of 1st issue so that we can analyze better solutions.

**Comment by vasujain275** [comment]
@kb-0311
1. Yes, the workflow is breaking because of the remote execution of shell commands in short intervals.
2. The root cause of the 1st Issue is, in my opinion, the time it takes to execute remote commands is too long. That's Why I think breaking the workflow into two can be a valid solution to both problems.

**Comment by kb-0311** [comment]
@vasujain275 sorry for the late reply. You can proceed with this approach. I would even say after the workflow to push the api to docker hub is executed, delay the execution of the workflow for pulling the API in VPS by some reasonable amount of time. This should solve our problem if SSH execution is the bottleneck.

**Comment by xoldyckk** [comment]
There's fly.io and render for web server hosting. Anyway, nowadays most companies have stopped providing good free tiers for stateful continously running servers, most free options are available only in serverless. But talawa-api isn't build around the idea of serverless.

Mongodb is not the only cloud database with free tier. Infact sql database providers have more generous free tiers compared to mongodb. For postgresql there's supabase and neondb, for mysql there's planetscale.

**Comment by kb-0311** [comment]
@xoldyckk The api has already been deployed on a VPS : https://api-demo.talawa.io/graphql

**Comment by vasujain275** [comment]
> @vasujain275 sorry for the late reply. You can proceed with this approach. I would even say after the workflow to push the api to docker hub is executed, delay the execution of the workflow for pulling the API in VPS by some reasonable amount of time. This should solve our problem if SSH execution is the bottleneck.

Ok, will complete this new workflow and add a PR for it

**Comment by vasujain275** [comment]
@palisadoes @kb-0311 @xoldyckk Here are the updates on this -

1. Waiting for @varshith257 to complete the multi-stage build Dockerfile here at - #1751.
2. Once the new multi-stage Dockerfile is merged, I will use my personal Docker Hub account to create and test the Talawa API GitHub Action for uploading images to Docker Hub on every push.
3. Once the pipeline is done and tested, we can add the official Palisadoes Foundation Docker Hub account credentials as GitHub Secrets and test the pipeline and cloud instance at - [https://api-demo.talawa.io/](https://api-demo.talawa.io/)

**Comment by kb-0311** [comment]
@vasujain275 That sounds great. 

@varshith257 Do keep me updated on the PR when you submit to implement multi stage Dockerfile.

**Comment by varshith257** [comment]
@kb-0311 I have done implementing multi-stage Dockerfile build and awaiting to make a PR but issue is not assigned yet to me. Waiting for approval from @palisadoes or @Cioppolo14 .

**Comment by Cioppolo14** [comment]
@varshith257 We are working on a solution right now for Docker Hub. We will follow up.  Thanks for your patience.

**Comment by vasujain275** [comment]
Update on this - 
The work on this issue is paused due to this issue being dependent on - #1885 being merged. Once this pr is merged, I will start work on it according to this approach - https://github.com/PalisadoesFoundation/talawa-api/issues/1428#issuecomment-1957799874

**Comment by vasujain275** [comment]
This issue is still stalled due to #1885 not being merged. The PR is waiting for review

**Comment by varshith257** [comment]
@vasujain275 If anything make you halt to make progress because of #1885. Please take work from my PR and test it and go ahead to progress.

@palisadoes is really worried about to see this ASAP to make available to all of us

**Comment by vasujain275** [comment]
@varshith257 Okay, I will take docker files from your pr and paste it in my fork to test the ci cd pipeline

**Comment by varshith257** [comment]
👍

**Comment by vasujain275** [comment]
> @varshith257 Okay, I will take docker files from your pr and paste it in my fork to test the ci cd pipeline

Sorry for no progress last week on this issue, I was not well. Resuming the work again

**Comment by varshith257** [comment]
@vasujain275 Are you fine now?

**Comment by vasujain275** [comment]
> @vasujain275 Are you fine now? 

Better now, thanks for asking

**Comment by varshith257** [comment]
Any progress on this?

**Comment by vasujain275** [comment]
> Any progress on this?

Yeah, I am researching a solution that will be effective than current one. Currently, we are using `npm run dev` on the demo api server containers. Does the build process issue you were working on is fixed? If it is resolved, we can then serve pre built html css js files directly.

**Comment by varshith257** [comment]
Yes the build issue is got fixed

**Comment by vasujain275** [comment]
PR #1751 by @varshith257  merged yesterday only, will work on this from today

**Comment by vasujain275** [comment]
@palisadoes Sorry for late reply, I am currently busy at school due to end sem exams till 30 May. I can continue the work after that If anybody do not solve this issue till them. Sorry for the inconvenience

**Comment by palisadoes** [comment]
@vasujain275 

1. Thanks. This can wait.
2. We have an issue created to fix the Docker data importation which should help when you return.

**Comment by vasujain275** [comment]
Thanks for understanding

**Comment by vasujain275** [comment]
@varshith257 Did you fix the build issues as I am not able to get the server up and running in production mode, here is the list of commands I run - 
```zsh
npm install
npm setup
npm run prebuild
npm run build
npm run start
```

Please take a look into it

**Comment by palisadoes** [comment]
@noman2002 @Cioppolo14 This is a cloud instance issue to follow.

**Comment by vasujain275** [comment]
I have reviewed the Issue that was raised to fix the build errors - https://github.com/PalisadoesFoundation/talawa-api/issues/1916 but saw that was not able to successfully fix all errors in the build process due to stability issues. 

@palisadoes shall we continue with dev images only as we are doing currently? They will 3x the size of production docker images (600MB -> 1.8GB) but this is the only viable option imo right now

**Comment by varshith257** [comment]
Why 1.8GB image size is required? The multi-stage build has been done to reduce image size, it is now ~500MB of dev and prod (~200MB).

**Comment by varshith257** [comment]
> @varshith257 Did you fix the build issues as I am not able to get the server up and running in production mode, here is the list of commands I run -

It won't because we are still pointing it to a stale file. There needs a refactoring of the codebase to support the production environment. I have closed the issue by fixing some of them, But it needs to be fixed with tsconfig and fixing of the errors emitted. The compiler from ts-> js is accompanied by incompatibility and it breaks down due to some libraries being built of ECMA.

**Comment by palisadoes** [comment]
1. We just need an instance up and running on the GoDaddy server.
2. It must work whenever we update the main / or develop branch. (I'm not sure which yet in my mind)
3. We are paying for a docker hub account that we are not using. I'd prefer to shut it down until we are ready. Can I do that?

**Comment by varshith257** [comment]
@vasujain275 You can go with testing it with dev image. As we are developing this for developers dev image could be right one to go with IMO

**Comment by vasujain275** [comment]
@palisadoes 
1. Okay, I will go with dev images only as the build process is not ready yet.
2. Yes CI/CD pipeline will ensure to update the instance on every new commit.
3. Why are we paying for a Docker Hub account?? I specifically mentioned that we don't need a docker hub paid account, please downgrade that account to a free account.

**Comment by palisadoes** [comment]
1. We had to use a paid account as we needed to register as an organization. 
2. Individuals can get free accounts, but we don't want to be tied to someone who could walk away from the project.

Do we need a docker account at this stage?

**Comment by vasujain275** [comment]
@palisadoes If we need a paid account for docker hub then I can implement the pipeline without docker hub. Docker hub would make this a little easier but we can implement the pipeline without it too.

**Comment by vasujain275** [comment]
@palisadoes 
The ci/cd is ready but it is failing often when there are several commits in small time intervals. How about if we update the cloud instance every 6 or 12 hours? It will improve the stability and uptime of the instance significantly.

**Comment by Cioppolo14** [comment]
> @palisadoes The ci/cd is ready but it is failing often when there are several commits in small time intervals. How about if we update the cloud instance every 6 or 12 hours? It will improve the stability and uptime of the instance significantly.

@noman2002 @palisadoes What do you think?

**Comment by palisadoes** [comment]
> @palisadoes The ci/cd is ready but it is failing often when there are several commits in small time intervals. How about if we update the cloud instance every 6 or 12 hours? It will improve the stability and uptime of the instance significantly.

Once a day, and try to make it happen with a reloading of the sample database. It'll be easier to troubleshoot with the lower frequency

**Comment by palisadoes** [comment]
Thanks for continuing with this

**Comment by vasujain275** [comment]
@palisadoes 

1. I will change the github action to run once every 24 hours. 
2. Caddy is a much better alternative to Nginx + Certbot. I will raise an issue to migrate to caddy.

**Comment by vasujain275** [comment]
@palisadoes As I am revamping the build and deploy process of talawa-api, I think it will be easy to start with fresh vps. So, can you please reset our current VPS that we are working on?

**Comment by palisadoes** [comment]
Why will a fresh VPS be easier?

**Comment by vasujain275** [comment]
I have added a lot of manual configurations and cron jobs to get Nginx Certbot working, but that won't be needed now because the Caddy reverse proxy only needs a Caddyfile and additions in Docker Compose and does not require additional manual configuration on the server. So, a fresh VPS will make it a lot easier to document the process. If resetting the VPS is not possible easily, then I can manage with this current vps state as well.

**Comment by vasujain275** [comment]
@palisadoes No need to reset now, I am reverting all the nginx cerbot configs by cli. Will move them to Caddy

**Comment by palisadoes** [comment]
Thanks.

**Comment by vasujain275** [comment]
As Caddy Pr was closed due to branch name issue, I will make another pr by tomorrow as @palisadoes suggested

**Comment by Cioppolo14** [comment]
@vasujain275 Should we see if someone else wants to pick this up or do you have time for this?

**Comment by vasujain275** [comment]
> @vasujain275 Should we see if someone else wants to pick this up or do you have time for this? 
> 

I can work on it but we are shifting from mongodb to Postgresql soon, so the only left feature in this issue is that the db sample data populate script will be obsolete in few weeks. 

The demo api deployment is working fine and has been hosted on the demo api website for ~10 weeks.

**Comment by vasujain275** [comment]
@palisadoes The Deploying to VPS workflow is failing because we decided not to deploy on each commit but instead deploy evert 24hrs. I haven't modify the github action to deploy on 24hr instead of every commit

**Comment by vasujain275** [comment]
@palisadoes Any update on database migration from mongodb to postgresql? Because I haven't made the sample data push script due db being migrated from mongo to postgresql

**Comment by palisadoes** [comment]
1. @xoldd has submitted a PR on a new branch
2. @vasujain275 we should assume the continued use of mongo until we migrate all the other repos to use the postgres option.
3. Take a look at the most recent post in the #general slack channel

**Comment by palisadoes** [comment]
@vasujain275 The postgres merge has happened. there is now a `develop-postgres` branch with the updates

**Comment by palisadoes** [comment]
@vasujain275 I was talking to @AVtheking and he feels we should launch this with the `develop` branch and figure out the `develop-postgres` work later. I can get `develop` up to date to assist with that process

**Comment by vasujain275** [comment]
@palisadoes Okay, I will test the docker files for MongoDb and make a pr. We can have the develop branch up and running for now.

**Comment by palisadoes** [comment]
> @palisadoes Okay, I will test the docker files for MongoDb and make a pr. We can have the develop branch up and running for now.

@vasujain275

**Comment by palisadoes** [comment]
@vasujain275, I was talking to @AVtheking who feels we also need a live instance of the `develop` branch of Talawa-Admin running too on the same server. This would help new users and developers to get a better feel for the application. Ideally we'd use the docker versions of the API and Admin.

The requirements would be:
1. Meet the original requirements of this issue
    1. https://github.com/PalisadoesFoundation/talawa-api/issues/1428#issue-1997921815
1. Create the cloud API instance such that the API listens on all interfaces on the default port. 
1. Merge the Talawa-Admin `develop-postgres` branch into `develop`. Have

**Comment by palisadoes** [comment]
@SiddheshKukade got an account on the server and has managed to get the API up and running. He's working on ensuring the scripts in the repo work.

**Comment by vasujain275** [comment]
@SiddheshKukade Let me know if you need any help with getting API / scripts setup on the VPS.

@palisadoes If we want to run both API and admin on the same VPS, I think we should add more compute and storage to it. I think it will bottleneck on the current tier if we run both api and admin on that instance.

**Comment by palisadoes** [comment]
> @SiddheshKukade Let me know if you need any help with getting API / scripts setup on the VPS.
> 
> @palisadoes If we want to run both API and admin on the same VPS, I think we should add more compute and storage to it. I think it will bottleneck on the current tier if we run both api and admin on that instance.

How many cores an how much RAM. We don't need a lot of disk

**Comment by palisadoes** [comment]
We have an issue to try to reduce the docker resource utilization

- https://github.com/PalisadoesFoundation/talawa-api/issues/2730#issuecomment-2513819481

**Comment by palisadoes** [comment]
> @SiddheshKukade Let me know if you need any help with getting API / scripts setup on the VPS.
> 
> @palisadoes If we want to run both API and admin on the same VPS, I think we should add more compute and storage to it. I think it will bottleneck on the current tier if we run both api and admin on that instance.

This is what we have:

**Comment by palisadoes** [comment]
> @SiddheshKukade Let me know if you need any help with getting API / scripts setup on the VPS.
> 
> @palisadoes If we want to run both API and admin on the same VPS, I think we should add more compute and storage to it. I think it will bottleneck on the current tier if we run both api and admin on that instance.

This is what we have, it can be 2X or 4X that, but I'd prefer to do the minimum upgrade.

![image](https://github.com/user-attachments/assets/c7830038-0634-41de-b9c3-1bd4e58dffdb)

**Comment by vasujain275** [comment]
@palisadoes I think we should go for 4v CPU Cores and 8 GB RAM. I will then test running both API and ADMIN on it simultaneously and can get back to you with the results.

**Comment by palisadoes** [comment]
@vasujain275 @SiddheshKukade The server has been upgraded to this:

![image](https://github.com/user-attachments/assets/a1c3cb2e-f56b-4cde-8f95-1521d6015ed3)

Please coordinate on getting this to work. Remember we want this to be completely automated:

1. Every 24 hours:
    1. The DB is reset
    2. The local repos are reset to the latest `develop`
3. The API and Admin instances must reload with every new PR merge into the `develop` branch

I'm hoping that this will be sufficient as an initial security measure. 

Additionally:

1. I have created a `talawa-admin` user for Admin to use when running
2. I have created a `talawa-API` user for API to use when running
3. Use Apache as a reverse proxy for both apps

When it is all working we'll need the posix shell for the talawa-* users to be `/usr/sbin/nologin` for added security. Please investigate how to get docker running as a user running with this shell. It's very important.

**Comment by palisadoes** [comment]
The DNS entries to use are:

1. demo.talawa.io (Admin)
2. api-demo.talawa.io (API)

They are the same server, but the separate DNS entries should help us in future if we want to host them on separate machines.

**Comment by palisadoes** [comment]
@SiddheshKukade The API didn't survive the machine reboot when it was upgraded.

**Comment by palisadoes** [comment]
Team,

You can use these files for guidance in setting up Admin and the API as linux daemons. They are from our switchmap repo and work correctly. This will allow you to control the apps using the `systemctl` command

- https://github.com/PalisadoesFoundation/switchmap-ng/tree/main/examples/linux/systemd

You can use this file as an configuration example for setting up an Apache reverse proxy. 

- https://github.com/PalisadoesFoundation/switchmap-ng/tree/main/examples/linux/apache

You can use this file as an configuration example for setting up an Nginx reverse proxy. 

- https://github.com/PalisadoesFoundation/switchmap-ng/tree/main/examples/linux/nginx

You'll have to adjust them accordingly.

Unless you can find a way to reduce the Docker resource utilization we'll need to run the apps natively.

**Comment by palisadoes** [comment]
I've created two issues that hopefully will help with the long term solution

- https://github.com/PalisadoesFoundation/talawa-api/issues/2737
- https://github.com/PalisadoesFoundation/talawa-admin/issues/2595

**Comment by vasujain275** [comment]
@palisadoes 

The Docker performance is extremely poor due to issues with our build process. This was the root cause earlier when we encountered a block, and it still is. We need to prioritize fixing the build process first.

To address this, we require someone with good GraphQL experience to resolve the build process issues. Once that is done, we can easily deploy the API admin on the same VPS using Docker alone. This approach will be highly efficient.

With Docker and Caddy (as the reverse proxy), there will be no need to create a systemd files too.

**Comment by palisadoes** [comment]
@vasujain275 Please create an issue detailing what is required. Hopefully we can find someone who will be willing to dedicate the time. In the issue, please state that we need an experienced person with a long and active GitHub history.

**Comment by palisadoes** [comment]
@SiddheshKukade We still need a running API instance. Have you managed to do more testing of your approach? I'd like the community to start using it before the end of the year

**Comment by vasujain275** [comment]
@palisadoes Okay let me create an issue with error logs and requirements

**Comment by vasujain275** [comment]
@palisadoes 

1. I have verifed that recent the recently merged pr by @adithyanotfound fixed the build issue. Now the `npm run prod` is working as aspected.
2. Now we can run the `talawa-api` and `talawa-admin` both on a single VPS using docker with the performance issues.
3. I am currently working on the modified docker files and will submit a pr with those and new scripts.
4. Here is the worflow I came with, listing it here to avoid confusion - 

- Both `talawa-api` and `talawa-admin` will run on `talawa` user with caddy serving as reverse proxy in both of there docker image. 
- By using Caddy instead of nginx as a revese proxy we elemimate the need for certbot and its realted scripts.
- Every 24hr a cron job will run that will pull the latest commits and rebuild the docker containers in the VPS
- Every 24hr another cron job will also run to reset the mongodb database.

Do you suggest any changes or should I proceed with this?

**Comment by palisadoes** [comment]
@vasujain275 

Based on this, do we need to differentiate between prod and dev environments? The poor performance of the dev environment for the API is a real cause of hardship. I'd prefer a great user experience for both end users and developers alike.

Maintaining two environments as an open source project with limited resources doesn't seem feasible to me. Is this a practical solution? Do you see any potential downsides to this approach, and if so do they outweigh the benefits in your opinion?

I welcome comments from others

**Comment by vasujain275** [comment]
@palisadoes 

We don't need two different environments for talawa api, the production one will be sufficient to handle both requests from talawa admin/app and developers.

Do we also need Apollo Graphql Playground Hosted on the VPS? Because it will be resource intensive as it requires self hosting a standalone apollo server. Currently the prod build disables the Apollo playground i.e. `/graphql` endpoint.

**Comment by adithyanotfound** [comment]
@palisadoes 

The production environment should be enough. Although having a development instance would be nice since we can then implement CD pipelines which would make deployment more easier. But due to limited resources, the best course of action would be to just have the prod instance. The current PR workflow should be sufficient to handle most cases. Occasionally, any errors can be addressed when the main branch is merged with the develop branch, if necessary.

**Comment by palisadoes** [comment]
1. Lets get rid of the development environment all together. I'll create an issue for it.
2. The resource intensity is too great in dev.

**Comment by vasujain275** [comment]
@palisadoes There is no need to remove dev docker files, developers can use that on their local machines for local development. We will use production docker files in cloud instance anyway.

**Comment by palisadoes** [comment]
Thanks

**Comment by palisadoes** [comment]
Here is the issue

- https://github.com/PalisadoesFoundation/talawa-api/issues/2800

**Comment by vasujain275** [comment]
@palisadoes 

I will in the meantime update the cloud instance scripts, production docker files and test them on `vasu` user on the vps.

**Comment by palisadoes** [comment]
This will be an interesting issue. Get ready for the migration!
- https://github.com/PalisadoesFoundation/talawa-api/issues/2808

**Comment by vasujain275** [comment]
@palisadoes 

Should I wait for this postgres migration to complete first? Or are we planning to go with mongodb for now?

**Comment by palisadoes** [comment]
1. We'll run everything on the server using `develop`. 
2. Just before the migration starts I'll merge `develop-postgres` into `develop`

**Comment by vasujain275** [comment]
@palisadoes 

Okay, I will wait for the merge then modify the docker files and scripts for postgres only.

**Comment by palisadoes** [comment]
@vasujain275 

1. We need this functioning. Only on `develop`.
1. I've just merged the latest Admin `develop-postgres` into `develop`
     1. https://github.com/PalisadoesFoundation/talawa-admin/pull/3183
1. Please try to get this running by the weekend.

**Comment by vasujain275** [comment]
@palisadoes 

I am free this weekend, will try to get this running by this monday

**Comment by palisadoes** [comment]
@vasujain275 

1. We only need the cloud instance for the `develop` branch at this time for both the API and Admin.
3. We are a long way away from having a working `develop-postgres` branch.

**Comment by palisadoes** [comment]
@vasujain275 

1. Were you able to make any progress?
2. Our deadline is to get this running before February 11 when our GSoC organization application is submitted.
3. Can you meet that deadline? We need to redouble our efforts.

**Comment by palisadoes** [comment]
1. I need a volunteer to take this issue over as @vasujain275 doesn't seem to be available.
    - https://github.com/PalisadoesFoundation/talawa-api/issues/1428
1. The API needs to run under the `talawa-api` user and Admin needs to run under the `talawa-admin` user 
1. The host is `api-demo.talawa.io` which is the same OS instance as `admin-demo.talawa.io`
2. Whoever is interested in working on this, contact me on slack to post their public SSH key. They will also get `sudo` access.
3. This absolutely needs to be resolved this weekend. We are highlighting the demo as part of our GSoC 2025 application, as justifiable validation of our progress.

**Comment by vasujain275** [comment]
@palisadoes Sorry for the inactivity and delayer response, due to personal reasons I am not able to give time to this issue and will not able to meet the GSOC deadline. Unassigning myself

**Comment by palisadoes** [comment]
@vasujain275 

I hope you persevere to overcome the recent challenges. Please stay in touch.

**Comment by hkumar1729** [comment]
I would like to continue working on this issue. I will coordinate with @noman2002 and @kb-0311 to finalize the setup. Please assign me and provide any necessary details to proceed. Thanks!

**Comment by PurnenduMIshra129th** [comment]
@palisadoes  i  want to work on this issue. please assign i will solve it.

**Comment by PurnenduMIshra129th** [comment]
@noman2002 @kb-0311 please assign i will try to solve it.

**Comment by palisadoes** [comment]
This is being worked on

**Comment by palisadoes** [comment]
I'm going to close this. It's mostly solved.
